pkg/assembler/backends/keyvalue/certifyBad.go

```diff
@@ -192,7 +193,171 @@ func (c *demoClient) ingestCertifyBad(ctx context.Context, subject model.Package
 // Query CertifyBad
 
 func (c *demoClient) CertifyBadList(ctx context.Context, certifyBadSpec model.CertifyBadSpec, after *string, first *int) (*model.CertifyBadConnection, error) {
-	return nil, fmt.Errorf("not implemented: CertifyBadList")
+	c.m.RLock()
+	defer c.m.RUnlock()
+
+	funcName := "CertifyBad"
+
+	if certifyBadSpec.ID != nil {
+		link, err := byIDkv[*badLink](ctx, *certifyBadSpec.ID, c)
+		if err != nil {
+			// Not found
+			return nil, nil
+		}
+		foundCertifyBad, err := c.buildCertifyBad(ctx, link, &certifyBadSpec, true)
+		if err != nil {
+			return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+		}
+		return &model.CertifyBadConnection{
+			TotalCount: 1,
+			PageInfo: &model.PageInfo{
+				HasNextPage: false,
+				StartCursor: ptrfrom.String(foundCertifyBad.ID),
+				EndCursor:   ptrfrom.String(foundCertifyBad.ID),
+			},
+			Edges: []*model.CertifyBadEdge{
+				{
+					Cursor: foundCertifyBad.ID,
+					Node:   foundCertifyBad,
+				},
+			},
+		}, nil
+	}
+
+	edges := make([]*model.CertifyBadEdge, 0)
+	hasNextPage := false
+	numNodes := 0
+	totalCount := 0
+	addToCount := 0
+
+	// Cant really search for an exact Pkg, as these can be linked to either
+	// names or versions, and version could be empty.
+	var search []string
+	foundOne := false
+	if certifyBadSpec.Subject != nil && certifyBadSpec.Subject.Artifact != nil {
+		exactArtifact, err := c.artifactExact(ctx, certifyBadSpec.Subject.Artifact)
+		if err != nil {
+			return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+		}
+		if exactArtifact != nil {
+			search = append(search, exactArtifact.BadLinks...)
+			foundOne = true
+		}
+	}
+	if !foundOne && certifyBadSpec.Subject != nil && certifyBadSpec.Subject.Source != nil {
+		exactSource, err := c.exactSource(ctx, certifyBadSpec.Subject.Source)
+		if err != nil {
+			return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+		}
+		if exactSource != nil {
+			search = append(search, exactSource.BadLinks...)
+			foundOne = true
+		}
+	}
+
+	if foundOne {
+		for _, id := range search {
+			link, err := byIDkv[*badLink](ctx, id, c)
+			if err != nil {
+				return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+			}
+			cb, err := c.CBIfMatch(ctx, &certifyBadSpec, link)
+			if err != nil {
+				return nil, gqlerror.Errorf("%v :: %v", funcName, err)
+			}
+			if cb == nil {
+				continue
+			}
+
+			if after != nil {
```

Comment:
need to set if there is a next page or not. Also is `currentPage` not needed here?

Reply:
Thanks, turns out that my logic for checking `after` was a little bit off. But we don’t need a `currentPage` variable because we are only checking for nodes that aren’t sorted. Since they aren’t sorted, we have to do a check like: `(after != nil && cb.ID > *after) || after == nil` for every iteration.